src/twisted/application/runner/_pidfile.py

```diff
@@ -0,0 +1,270 @@
+        """
+
+
+    def write(pid):
+        """
+        Store a PID in this PID file.
+
+        @param pid: A PID to store.
+        @type pid: L{int}
+
+        @raise EnvironmentError: If this PID file cannot be written.
+        """
+
+
+    def writeRunningPID():
+        """
+        Store the PID of the current process in this PID file.
+
+        @raise EnvironmentError: If this PID file cannot be written.
+        """
+
+
+    def remove():
+        """
+        Remove this PID file.
+
+        @raise EnvironmentError: If this PID file cannot be removed.
+        """
+
+
+    def isRunning():
+        """
+        Determine whether there is a running process corresponding to the PID
+        in this PID file.
+
+        @return: True if this PID file contains a PID and a process with that
+        PID is currently running; false otherwise.
+        @rtype: L{bool}
+
+        @raise EnvironmentError: If this PID file cannot be read.
+        @raise InvalidPIDFileError: If this PID file's content is invalid.
+        @raise StalePIDFileError: If this PID file's content refers to a PID
+            for which there is no corresponding running process.
+        """
+
+
+    def __enter__():
+        """
+        Enter a context using this PIDFile.
+
+        Writes the PID file with the PID of the running process.
+
+        @raise AlreadyRunningError: A process corresponding to the PID in this
+            PID file is already running.
+        """
+
+
+    def __exit__(exc_type, exc_value, traceback):
+        """
+        Exit a context using this PIDFile.
+
+        Removes the PID file.
+        """
+
+
+
+@implementer(IPIDFile)
+class PIDFile(object):
+    """
+    Concrete implementation of L{IPIDFile} based on C{IFilePath}.
+    """
+
+    _log = Logger()
+
+
+    @staticmethod
+    def format(pid):
+        """
+        Format a PID file's content.
+
+        @param pid: A process ID.
+        @type pid: int
+
+        @return: Formatted PID file contents.
+        @rtype: L{bytes}
+        """
+        return u"{}\n".format(int(pid)).encode("utf-8")
+
+
+    def __init__(self, filePath):
+        """
+        @param filePath: The path to the PID file on disk.
+        @type filePath: L{IFilePath}
+        """
+        self.filePath = filePath
+
+
+    def read(self):
+        pidString = b""
+        try:
+            with self.filePath.open() as fh:
+                for pidString in fh:
+                    break
+        except OSError as e:
+            if e.errno == errno.ENOENT:  # No such file
+                raise NoPIDFound("PID file does not exist")
+            raise
+
+        try:
+            return int(pidString)
+        except ValueError:
+            raise InvalidPIDFileError(
+                "non-integer PID value in PID file: {!r}".format(pidString)
+            )
+
+
+    def write(self, pid):
+        self.filePath.setContent(self.format(pid=pid))
```

Comment:
This is a race-condition-prone implementation mechanism.

If two `twist` processes start up simultaneously, one should win the race, and one should lose; a pidfile is a sort of lock file.  With this strategy, it's look-before-you-leap with all the attendant problems that might cause.

The only really _right_ way to create a "file" with "contents" atomically is to use a symlink, which is what `twisted.python.lockfile.FilesystemLock` does.  However, if we want to preserve the existing contract of a pidfile being a literal file with bytes in it and not a symlink, we can use `FilePath.create` instead.

My concern with this is that I don't want a public `.write` method floating around with these _semantics_; I'm OK with the bug existing as long as a follow-up PR can fix it without disturbing the new public interface.